File: apps/anila-shell/src/runtime/continueSeam.js
// 續寫接到同一則答案上。
// 只有整段重複（至少 40 字）才剪掉續寫開頭。短的相同字尾，像程式裡第二個
// return，留著，避免改掉語意。
// 程式碼圍欄沒關時不剪重疊、也不插空白；只拿掉確實重複的那道開頭。
// 一般文字若在英數中間被截斷，接縫補一個空白。

const MIN_OVERLAP = 40;
const FENCE_LINE = /^( {0,3})(`{3,}|~{3,})(.*)$/;

/** 跟 Router 約定的同一句。續寫回合由 Shell 放進使用者訊息。 */
export const CONTINUE_INSTRUCTION =
  "請接續上文，直接從中斷處往下寫，不要重複已經寫過的內容。";

function fenceMarker(line) {
  const matched = FENCE_LINE.exec(line);
  if (!matched) return null;
  const tick = matched[2];
  const info = matched[3] ?? "";
  if (tick[0] === "`" && info.includes("`")) return null;
  return { char: tick[0], len: tick.length, info };
}

function unclosedFence(text) {
  let open = null;
  for (const line of String(text || "").split("\n")) {
    const marker = fenceMarker(line);
    if (!marker) continue;
    if (
      open
      && marker.char === open.char
      && marker.len >= open.len
      && marker.info.trim() === ""
    ) {
      open = null;
      continue;
    }
    if (!open) open = marker;
  }
  return open;
}

function stripRepeatedOpener(extra, open) {
  const lines = extra.split("\n");
  let index = 0;
  while (index < lines.length && lines[index].trim() === "") index += 1;
  if (index >= lines.length) return extra;
  const marker = fenceMarker(lines[index]);
  if (!marker || marker.char !== open.char) return extra;
  // 長度夠、後面沒有資訊，才是 CommonMark 的關閉標記，要留著。
  const closes = marker.len >= open.len && marker.info.trim() === "";
  if (closes) return extra;
  // 字元、長度、資訊都一樣，才是把開頭再寫一次。較短的圍欄是內容。
  const sameOpener = marker.len === open.len && marker.info.trim() === open.info.trim();
  if (!sameOpener) return extra;
  lines.splice(0, index + 1);
  return lines.join("\n");
}

function overlapLength(prior, extra) {
  const max = Math.min(prior.length, extra.length, 400);
  for (let size = max; size >= MIN_OVERLAP; size -= 1) {
    if (prior.slice(-size) === extra.slice(0, size)) return size;
  }
  return 0;
}

function dropOverlap(prior, extra) {
  const direct = overlapLength(prior, extra);
  if (direct > 0) return extra.slice(direct);
  const stripped = extra.replace(/^\n+/, "");
  if (stripped === extra) return extra;
  const nested = overlapLength(prior, stripped);
  if (nested > 0) return stripped.slice(nested);
  return extra;
}

function boundarySpace(base, extra) {
  const left = base.slice(-1);
  const right = extra.slice(0, 1);
  if (!left || !right) return "";
  if (/\s/.test(left) || /\s/.test(right)) return "";
  if (/[A-Za-z0-9]/.test(left) && /[A-Za-z0-9]/.test(right)) return " ";
  return "";
}

/** 把這一輪續寫接到已寫出的正文後面。空的續寫就留原文。 */
export function joinContinuation(prior, addition) {
  const base = typeof prior === "string" ? prior : "";
  let extra = typeof addition === "string" ? addition : "";
  if (!extra) return base;
  if (!base) return extra;
  const open = unclosedFence(base);
  if (open) extra = stripRepeatedOpener(extra, open);
  // 圍欄裡的重疊分不出是重播還是下一行真的要同樣的程式，不剪。
  if (!open) extra = dropOverlap(base, extra);
  if (!extra) return base;
  const gap = open ? "" : boundarySpace(base, extra);
  return base + gap + extra;
}

File: packages/anila-core/src/anila_core/api/continue_seam.py
"""把下一輪接到已經送出的答案後面。

跟 Shell 的 continueSeam 同一套：只有整段重複（至少 40 字）才剪掉
續寫開頭。程式碼圍欄沒關時不剪重疊、也不插空白。
標記自己佔一行，所以重疊比對會再試一次拿掉結尾換行的前文。
"""

from __future__ import annotations

import re

_MIN_OVERLAP = 40
_FENCE_LINE = re.compile(r"^( {0,3})(`{3,}|~{3,})(.*)$")


def _fence_marker(line: str) -> dict[str, object] | None:
    matched = _FENCE_LINE.match(line)
    if matched is None:
        return None
    tick = matched.group(2)
    info = matched.group(3) or ""
    if tick[0] == "`" and "`" in info:
        return None
    return {"char": tick[0], "length": len(tick), "info": info}


def _unclosed_fence(text: str) -> dict[str, object] | None:
    open_fence: dict[str, object] | None = None
    for line in str(text or "").split("\n"):
        marker = _fence_marker(line)
        if marker is None:
            continue
        if (
            open_fence is not None
            and marker["char"] == open_fence["char"]
            and int(marker["length"]) >= int(open_fence["length"])
            and str(marker["info"]).strip() == ""
        ):
            open_fence = None
            continue
        if open_fence is None:
            open_fence = marker
    return open_fence


def _strip_repeated_opener(extra: str, open_fence: dict[str, object]) -> str:
    lines = extra.split("\n")
    index = 0
    while index < len(lines) and lines[index].strip() == "":
        index += 1
    if index >= len(lines):
        return extra
    marker = _fence_marker(lines[index])
    if marker is None or marker["char"] != open_fence["char"]:
        return extra
    closes = (
        int(marker["length"]) >= int(open_fence["length"])
        and str(marker["info"]).strip() == ""
    )
    if closes:
        return extra
    same = (
        int(marker["length"]) == int(open_fence["length"])
        and str(marker["info"]).strip() == str(open_fence["info"]).strip()
    )
    if not same:
        return extra
    del lines[: index + 1]
    return "\n".join(lines)


def _overlap_length(prior: str, extra: str) -> int:
    limit = min(len(prior), len(extra), 400)
    for size in range(limit, _MIN_OVERLAP - 1, -1):
        if prior[-size:] == extra[:size]:
            return size
    return 0


def _drop_overlap(prior: str, extra: str) -> str:
    direct = _overlap_length(prior, extra)
    if direct > 0:
        return extra[direct:]
    stripped = re.sub(r"^\n+", "", extra)
    if stripped == extra:
        return extra
    nested = _overlap_length(prior, stripped)
    if nested > 0:
        return stripped[nested:]
    return extra


def _boundary_space(base: str, extra: str) -> str:
    left = base[-1:] if base else ""
    right = extra[:1] if extra else ""
    if not left or not right:
        return ""
    if left.isspace() or right.isspace():
        return ""
    if re.match(r"[A-Za-z0-9]", left) and re.match(r"[A-Za-z0-9]", right):
        return " "
    return ""


def join_continuation(prior: str, addition: str) -> str:
    """跟 Shell ``joinContinuation`` 相同的接縫。"""
    base = prior if isinstance(prior, str) else ""
    extra = addition if isinstance(addition, str) else ""
    if not extra:
        return base
    if not base:
        return extra
    open_fence = _unclosed_fence(base)
    if open_fence is not None:
        extra = _strip_repeated_opener(extra, open_fence)
    if open_fence is None:
        extra = _drop_overlap(base, extra)
    if not extra:
        return base
    gap = "" if open_fence is not None else _boundary_space(base, extra)
    return base + gap + extra


def continuation_suffix(prior: str, addition: str) -> str:
    """前文已經送出。只回傳該接在後面的那一段。

    標記獨佔最後一行，前文常以換行作收。重疊若卡在那個換行前，
    就改用拿掉換行的前文來剪，已送出的換行留著。
    """
    extra = addition if isinstance(addition, str) else ""
    base = prior if isinstance(prior, str) else ""
    if not extra:
        return ""
    if not base:
        return extra
    direct = join_continuation(base, extra)
    suffix = direct[len(base) :] if direct.startswith(base) else extra
    if base.endswith("\n"):
        trimmed = base[:-1]
        alt = join_continuation(trimmed, extra)
        if alt.startswith(trimmed) and len(alt) < len(trimmed) + len(extra):
            return alt[len(trimmed) :]
    return suffix


def append_round(prior: str, addition: str) -> str:
    """把下一輪接到同一則答案。空的一輪就留原文。"""
    base = prior if isinstance(prior, str) else ""
    extra = addition if isinstance(addition, str) else ""
    if not base:
        return extra
    if not extra:
        return base
    return base + continuation_suffix(base, extra)


class ContinuationSeam:
    """串流時先留住開頭，重疊剪掉之後才送出。之後的字原樣接上。"""

    def __init__(self, prior: str) -> None:
        self.prior = prior if isinstance(prior, str) else ""
        self.buf = ""
        self.done = False

    def feed(self, chunk: str) -> str:
        if self.done:
            return chunk
        if chunk:
            self.buf += chunk
        if len(self.buf) < 400:
            return ""
        return self._release()

    def finish(self) -> str:
        if self.done:
            return ""
        return self._release()

    def _release(self) -> str:
        self.done = True
        addition = self.buf
        self.buf = ""
        return continuation_suffix(self.prior, addition)
